### PR TITLE
Bug 1848605 - Implement similar functionality as loadIntoView for Compose Top Sites

### PR DESCRIPTION
**Pull request by t-p-white**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1848605

### CODE REVIEW

File: android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt [hunk 1]
```diff
@@ -255,6 +262,47 @@ class BrowserIcons constructor(
         }
     }
 
+    @Composable
+    fun LoadIntoViewInternal(
```
Comment (t-p-white) [inline android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt:266]: this fun is essentially a hybrid of the `loadIntoViewInternal` fun above that is used in non-Compose Top Sites and the `BrowserIcons.Loader` that was used to previously load the icon by FavIcon

**Comment by t-p-white** [comment]
@gabrielluong @MozillaNoah interested to get your thoughts on this. Updating FavIcon to use the `LoadIntoViewInternal` will affect more than just Top Sites, however it seems to me that they will be updated to all be consistent

File: android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt (hunk 1 above)
Comment (MozillaNoah) [inline android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt:266]: Let's rename this to ditch the `View` and make it more clear that this is Composable UI and not just a business logic function with the `@Composable` tag.

```suggestion
    fun LoadableImage(
```

**Review by MozillaNoah — COMMENTED** [review]
I think this is looking good. I briefly tested this locally with synced tabs (as those utilize `Favicon` via `FaviconListItem`, and it seemed to be running well and perhaps a bit better than before.

Nice work!

**Comment by jonalmeida** [comment]
tenor-2801692567

File: android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt [hunk 2]
```diff
@@ -255,6 +262,57 @@ class BrowserIcons constructor(
...
+     * Loads an icon using [BrowserIcons] into the given Composable [content]. Synchronous loading
+     * via an in-memory cache is attempted first, followed by an asynchronous load as a fallback.
+     *
+     * @param url The URL of the website an icon should be loaded for.
+     * @param iconSize The preferred size of the icon that should be loaded.
+     * @param isPrivate Whether this request for this icon came from a private session.
+     * @param content The Composable content block to render the icon.
+     */
+    @Composable
+    fun LoadableImage(
```
Comment (MozillaNoah) [inline android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt:275]: does this need to still live inside the `BrowserIcons` class? it'll make it more encapsulated as its own thing but also clean up the usage inside of `Favicon`

**Review by MozillaNoah — APPROVED** [review]
LGTM. Just one question about potentially moving a Composable function outside of a class.

**Review by titooan — APPROVED** [review]
👏

File: android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt (hunk 2 above)
Comment (t-p-white) [reply android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt:275]: @MozillaNoah my reasoning to keep it in this inside the `BrowserIcons` class is that the fun is using a number of functions provided in the class:

- `private fun desiredSizeForRequest`
- `private fun loadIconMemoryOnly`
- `internal fun loadIconInternalAsync`

By keeping this in the same file it means fewer changes to the existing code which is my preference, wdyt?

Comment (MozillaNoah) [reply android-components/components/browser/icons/src/main/java/mozilla/components/browser/icons/BrowserIcons.kt:275]: ahh gotcha gotcha. I totally missed that several class functions were being used here. I think in an ideal world we'd have the Composable in its own file, decoupled from some of the specific nuances of `BrowserIcons`, but that is definitely way out of scope here, having given a full pass on that class.

This is good to go 👍

**Pull request merged**